plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/server/behavior/OpenShiftPublishController.java

```diff
@@ -186,4 +199,33 @@ protected boolean isEapProfile() {
 		String profile = ServerProfileModel.getProfile(getServer());
 		return OpenshiftEapProfileDetector.PROFILE.equals(profile);
 	}
+
+	public void deleteOldDeployments(final IServer server) throws CoreException {
+		File deploymentsRootFolder = getDeploymentsRootFolder();
+		Arrays.stream(server.getModules())
+			.forEach(module -> deleteOldDeployment(module, deploymentsRootFolder));
+	}
+
+	private void deleteOldDeployment(final IModule module, final File deploymentsRootFolder) {
+		OpenShiftServer delegate = OpenShiftServerUtils.getServerDelegate(getServer());
+		ModuleOutputNames moduleOutputNames = delegate.getModuleOutputNames();
+		if (moduleOutputNames.isModified(module)) {
+			deleteOldDeployment(module, moduleOutputNames, deploymentsRootFolder);
+			delegate.getModuleOutputNames().reset();
```

Comment:
This seems like a bug. If you're iterating through deleteOldDeployments(etc) and each loop then calling deleteOldDeployment, what happens if 5 deployments have changed (Wonka1.war through Wonka5.war)?   On the first iteration, you see Wonka1.war was modified. You reset the output names. On the second iteration, Wonka2 is marked as unchanged, because you've already reset the output names. It would have been marked as changed if you didn't just reset the cache on the previous iteration.

Reply:
true, should reset each module on it's own, not gloabbly.